### DISCUSSION TITLE
Allow iterating through select field options in Profile Edit form

### DISCUSSION THREAD

**Discussion by jHards** [ExpressionEngine/ExpressionEngine#4186] [Feature Requests]
**Description of the problem**
From discussion [here](https://github.com/ExpressionEngine/ExpressionEngine/pull/2804), it was anticipated that from 7.4, the code below would work. The text fields work fine, but the select fields do not. Docs are a little light on content on the edit profile tag, so I am making assumptions that it was indeed set-up in line with the Channel Form tag docs.

Text work ok, select does not.

**How To Reproduce**
```
{custom_profile_fields}
  {if '{field_id}' == 'x' OR '{field_id}' == 'y'}
    <label for="{field_name}" class="control-label">{field_label}:</label>
    {if '{field_type}' == 'select'}
      <select class="form-control" id="{field_name}" name="{field_name}">
        {options}<option value="{option_value}"{selected}>{option_name}</option>{/options}
      </select>
    {if:elseif '{field_type}' == 'text'}
      <span>{field_instructions}</span>
      <input type="text" class="form-control" dir="{text_direction}" id="{field_name}" name="{field_name}" value="{field_data}" maxlength="{maxlength}">
    {/if}
  {/if}
{/custom_profile_fields}
```

**Error Messages**
`{form:custom_profile_field}` gives no option for formatting:
![image](https://github.com/ExpressionEngine/ExpressionEngine/assets/17785714/db58493a-b427-46e1-9b99-409797c6e8a9)

The code above doesn't pull in the options:
![image](https://github.com/ExpressionEngine/ExpressionEngine/assets/17785714/1e029aa1-7dd1-400b-9011-0c6c884a3eed)

**Screenshots / Videos / Template Code**
N/A

**Environment Details:**
 - Version: 7.4.2

**Possible Solution**
N/A

**Additional context**
N/A

**Comment by intoeetive** [comment]
@jHards for the "formatting" issue, you would need to add [`include_assets`](https://docs.expressionengine.com/latest/member/edit-profile.html#include_assets) parameter to the form. You can also [pass the form parameters](https://docs.expressionengine.com/latest/member/edit-profile.html#parameters) such as class or data-attributes to the form

The `{options}` variable is not available. 

You can display the field withing `{custom_profile_fields}` loop using `{form:custom_profile_field}` tag or outside of the loop by referencing it by name `{field:my_select}`

https://docs.expressionengine.com/latest/member/edit-profile.html#example

Hope that helps

Again, if you have a specific need that can't be covered with these tags, please add a feature request

**Comment by jHards** [comment]
Thanks @intoeetive, I have the `include_assets` parameter at the form level, my assumption was that that would apply styling to the form, not to the underlying inputs? Here I am trying to add `class="form-control"` to the `<select>` tags. If `{options}` is unavailable, how would I do this?

**Comment by intoeetive** [comment]
I don't think you can do this. You can only add a wrapper div and add class to it. Ability to pass class name to input using `field:...` syntax would probably make a good feature request but it's not something we support currently

**Comment by jHards** [comment]
I gather from [the initial discussion](https://github.com/ExpressionEngine/ExpressionEngine/pull/2804) we had that you weren't planning on allowing a generic method to add a classes to input parameters, but I had hoped that we'd be able to build the inputs ourselves in line with [Channel Forms](https://docs.expressionengine.com/latest/channels/channel-form/fields.html#select-field) - impossible without `{options}{/options}` tags. Is this something I can support development on? Conscious that we've been waiting 9 months to see how this feature would actually ship in 7.4 and it's critical for our use case.

**Comment by intoeetive** [comment]
So we do have this [feature request](https://github.com/ExpressionEngine/ExpressionEngine/discussions/2870) (that you can a comment to, if you like) and [this PR](https://github.com/ExpressionEngine/ExpressionEngine/pull/3809) for it, but that's currently limited to just Channel Fields. That would allow you to do things like this:
```
    {exp:channel:field field_name="department"}
        <p>Choose <b>{field_label}</b>:</p>
            {field_options}
                <p><span class="value">{value}</span>: <span class="label">{label}</span></p>
            {/field_options}
        </div>
    {/exp:channel:field}
```
Now that we have Member Fields, it does probably make sense to add those as well - with same or a similar one.
What do you think?

**Comment by jHards** [comment]
I think this is the right sort of solution, but applied also to the Member fields. If I'm understanding correctly, it'll remove the need to use the current `{custom_profile_fields}{/custom_profile_fields}` and instead call each field in turn? With variables in-line with the manually constructed examples [here](https://docs.expressionengine.com/latest/channels/channel-form/fields.html#manually-constructed-input-fields)?

**Comment by intoeetive** [comment]
You already can display the individual field outside of `{custom_profile_fields}` loop by using `{field:my_field_name}` syntax https://docs.expressionengine.com/latest/member/edit-profile.html#custom-field

You normally need to construct the field completely manually if you need it to be different HTML element or something too custom

**Comment by jHards** [comment]
Unfortunately, as described above, the `{field:my_field_name}` tag doesn't allow for customization of the output, hence the attempt to manually constuct. In this example we are using [Bootstrap framework](https://getbootstrap.com/docs/4.0/components/forms/#form-controls) out of the box so need to include `class="form-control"` at the `<select>` or `<input>` tag level.

I'm unclear on the approach in #3809 - is this designed to work within the `{exp:channel:form}` tags or standalone? I'm expecting to not need to construct within the `{custom_profile_fields}{/custom_profile_fields}` and that my code within `{exp:member:edit_profile}` tags would become:

```
{field field_name="my_select_field"}
  <label for="{field_name}" class="control-label">{field_label}:</label>
  <select class="form-control" id="{field_name}" name="{field_name}">
    {options}<option value="{option_value}"{selected}>{option_name}</option>{/options}
  </select>
{/field}
{field field_name="my_text_field"}
  <label for="{field_name}" class="control-label">{field_label}:</label>
  <span>{field_instructions}</span>
  <input type="text" class="form-control" dir="{text_direction}" id="{field_name}" name="{field_name}" value="{field_data}" maxlength="{maxlength}">
{/field}
```

**Comment by intoeetive** [comment]
I will reopen and convert this into feature request. Thanks!

**Comment by jHards** [comment]
See [#4390](https://github.com/ExpressionEngine/ExpressionEngine/pull/4390)
